File: algorithms/dp/lcs.js
// =====================================================================
// LCS / Longest Common Subsequence / Наибольшая общая подпоследовательность
// ---------------------------------------------------------------------
// Подпоследовательность ≠ подстрока. "abc" — подпоследовательность "axbycz".
//
// dp[i][j] = LCS первых i символов A и первых j символов B
//   - A[i-1] === B[j-1]: dp[i][j] = dp[i-1][j-1] + 1
//   - иначе:             dp[i][j] = max(dp[i-1][j], dp[i][j-1])
//
// Time:  O(n * m)
// Space: O(n * m). Можно O(min(n, m)) с rolling rows.
//
// Применение: diff утилиты, биоинформатика (выравнивание последовательностей).
// =====================================================================

function lcsLength(a, b) {
    const n = a.length;
    const m = b.length;
    const dp = Array.from({ length: n + 1 }, () => new Array(m + 1).fill(0));

    for (let i = 1; i <= n; i++) {
        for (let j = 1; j <= m; j++) {
            if (a[i - 1] === b[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    return dp[n][m];
}

// Восстановление самой подпоследовательности
function lcs(a, b) {
    const n = a.length;
    const m = b.length;
    const dp = Array.from({ length: n + 1 }, () => new Array(m + 1).fill(0));

    for (let i = 1; i <= n; i++) {
        for (let j = 1; j <= m; j++) {
            dp[i][j] =
                a[i - 1] === b[j - 1]
                    ? dp[i - 1][j - 1] + 1
                    : Math.max(dp[i - 1][j], dp[i][j - 1]);
        }
    }

    // Идём от правого нижнего угла назад
    let i = n;
    let j = m;
    const chars = [];
    while (i > 0 && j > 0) {
        if (a[i - 1] === b[j - 1]) {
            chars.push(a[i - 1]);
            i--;
            j--;
        } else if (dp[i - 1][j] > dp[i][j - 1]) {
            i--;
        } else {
            j--;
        }
    }

    return chars.reverse().join('');
}

console.log(lcsLength('ABCBDAB', 'BDCAB')); // 4
console.log(lcs('ABCBDAB', 'BDCAB')); // 'BCAB' (один из вариантов)

module.exports = { lcsLength, lcs };

File: algorithms/dp/lcs.py
"""
=====================================================================
LCS / Longest Common Subsequence
---------------------------------------------------------------------
dp[i][j] = LCS A[:i], B[:j]
Time:  O(n*m), Space: O(n*m)
=====================================================================
"""


def lcs_length(a: str, b: str) -> int:
    n, m = len(a), len(b)
    dp = [[0] * (m + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if a[i - 1] == b[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    return dp[n][m]


def lcs(a: str, b: str) -> str:
    n, m = len(a), len(b)
    dp = [[0] * (m + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            dp[i][j] = (dp[i - 1][j - 1] + 1
                       if a[i - 1] == b[j - 1]
                       else max(dp[i - 1][j], dp[i][j - 1]))

    # Восстановление
    i, j = n, m
    chars = []
    while i > 0 and j > 0:
        if a[i - 1] == b[j - 1]:
            chars.append(a[i - 1])
            i -= 1
            j -= 1
        elif dp[i - 1][j] > dp[i][j - 1]:
            i -= 1
        else:
            j -= 1

    return "".join(reversed(chars))


if __name__ == "__main__":
    print(lcs_length("ABCBDAB", "BDCAB"))  # 4
    print(lcs("ABCBDAB", "BDCAB"))         # 'BCAB'
